File: client/src/main/java/au/com/codeka/warworlds/client/game/fleets/FleetsLayout.java
package au.com.codeka.warworlds.client.game.fleets;

import android.content.Context;
import android.transition.TransitionManager;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ExpandableListView;
import android.widget.FrameLayout;
import android.widget.RelativeLayout;

import au.com.codeka.warworlds.client.MainActivity;
import au.com.codeka.warworlds.client.R;
import au.com.codeka.warworlds.client.game.starfield.StarfieldManager;
import au.com.codeka.warworlds.client.game.world.StarCollection;
import au.com.codeka.warworlds.common.proto.Star;

/**
 * Layout for the {@link FleetsScreen}.
 */
public class FleetsLayout extends RelativeLayout {
  private final StarCollection starCollection;
  private final ExpandableListView listView;
  private final FleetExpandableStarListAdapter adapter;
  private final FrameLayout bottomPane;

  public FleetsLayout(Context context, StarCollection starCollection) {
    super(context);
    inflate(context, R.layout.fleets, this);
    this.starCollection = starCollection;

    bottomPane = findViewById(R.id.bottom_pane);

    listView = findViewById(R.id.fleet_list);
    adapter = new FleetExpandableStarListAdapter(LayoutInflater.from(getContext()), starCollection);
    listView.setAdapter(adapter);

    if (starCollection.size() == 1) {
      // if it's just one star, just expand it now.
      listView.expandGroup(0);
    }

    listView.setOnChildClickListener((lv, v, groupPosition, childPosition, id) -> {
      adapter.onItemClick(groupPosition, childPosition);
      return false;
    });

    // Actions pane by default.
    bottomPane.addView(new ActionBottomPane(getContext(), actionBottomPaneCallback));
  }

  /**
   * Select the given fleet. This is expensive and should be avoided except when there's only
   * one (or a small, finite number of) star.
   */
  public void selectFleet(long fleetId) {
    for (int groupPosition = 0; groupPosition < starCollection.size(); groupPosition++) {
      Star star = starCollection.get(groupPosition);
      listView.expandGroup(groupPosition);
      adapter.setSelectedFleetId(star, fleetId);
      break;
    }
  }

  @Override
  public void onDetachedFromWindow() {
    super.onDetachedFromWindow();

    adapter.destroy();
  }

  public void showActionsPane() {
    showStarfield(false /* visible */);

    TransitionManager.beginDelayedTransition(bottomPane);
    bottomPane.removeAllViews();
    bottomPane.addView(new ActionBottomPane(getContext(), actionBottomPaneCallback));
  }

  public void showMovePane(Star star, long fleetId) {
    showStarfield(true /* visible */);

    // TODO: the cast seems... not great.
    StarfieldManager starfieldManager = ((MainActivity) getContext()).getStarfieldManager();

    MoveBottomPane moveBottomPane =
        new MoveBottomPane(getContext(), starfieldManager, this::showActionsPane);
    moveBottomPane.setFleet(star, fleetId);

    TransitionManager.beginDelayedTransition(bottomPane);
    bottomPane.removeAllViews();
    bottomPane.addView(moveBottomPane);
  }

  public void showSplitPane(Star star, long fleetId) {
    showStarfield(false /* visible */);

    SplitBottomPane splitBottomPane = new SplitBottomPane(getContext(), this::showActionsPane);
    splitBottomPane.setFleet(star, fleetId);

    TransitionManager.beginDelayedTransition(bottomPane);
    bottomPane.removeAllViews();
    bottomPane.addView(splitBottomPane);
  }

  public void showMergePane(Star star, long fleetId) {
    showStarfield(false /* visible */);

    MergeBottomPane mergeBottomPane =
        new MergeBottomPane(getContext(), adapter, this::showActionsPane);
    mergeBottomPane.setFleet(star, fleetId);

    TransitionManager.beginDelayedTransition(bottomPane);
    bottomPane.removeAllViews();
    bottomPane.addView(mergeBottomPane);
  }

  public void showStarfield(boolean visible) {
    if (visible) {
      listView.setVisibility(View.GONE);
      setBackground(null);
    } else {
      listView.setVisibility(View.VISIBLE);
      setBackgroundResource(R.color.default_background);
    }
  }

  private final ActionBottomPane.Callback actionBottomPaneCallback =
      new ActionBottomPane.Callback() {
        @Override
        public void onMoveClick() {
          Long fleetId = adapter.getSelectedFleetId();
          if (fleetId == null) {
            return;
          }
          Star star = adapter.getSelectedStar();
          if (star == null) {
            return;
          }

          showMovePane(star, fleetId);
        }

        @Override
        public void onSplitClick() {
          Long fleetId = adapter.getSelectedFleetId();
          if (fleetId == null) {
            return;
          }
          Star star = adapter.getSelectedStar();
          if (star == null) {
            return;
          }

          showSplitPane(star, fleetId);
        }

        @Override
        public void onMergeClick() {
          Long fleetId = adapter.getSelectedFleetId();
          if (fleetId == null) {
            return;
          }
          Star star = adapter.getSelectedStar();
          if (star == null) {
            return;
          }

          showMergePane(star, fleetId);
        }
      };
}

File: client/src/main/kotlin/au/com/codeka/warworlds/client/game/fleets/FleetsLayout.kt
package au.com.codeka.warworlds.client.game.fleets

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.ExpandableListView
import android.widget.FrameLayout
import android.widget.RelativeLayout
import androidx.transition.TransitionManager
import au.com.codeka.warworlds.client.MainActivity
import au.com.codeka.warworlds.client.R
import au.com.codeka.warworlds.client.game.world.StarCollection
import au.com.codeka.warworlds.common.proto.Star
import au.com.codeka.warworlds.common.sim.FleetHelper

/** Layout for the [FleetsScreen]. */
class FleetsLayout(context: Context?, starCollection: StarCollection?) : RelativeLayout(context) {
  private val starCollection: StarCollection?
  private val listView: ExpandableListView
  private val adapter = FleetExpandableStarListAdapter(LayoutInflater.from(context), starCollection)
  private val bottomPane: FrameLayout

  /**
   * Select the given fleet. This is expensive and should be avoided except when there's only
   * one (or a small, finite number of) stars.
   */
  fun selectFleet(fleetId: Long) {
    for (groupPosition in 0 until starCollection!!.size()) {
      val star = starCollection[groupPosition]
      if (FleetHelper.findFleet(star, fleetId) != null) {
        listView.expandGroup(groupPosition)
        adapter.setSelectedFleetId(star, fleetId)
        break
      }
    }
  }

  public override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    adapter.destroy()
  }

  fun showActionsPane() {
    showStarfield(false /* visible */)
    TransitionManager.beginDelayedTransition(bottomPane)
    bottomPane.removeAllViews()
    bottomPane.addView(ActionBottomPane(context, actionBottomPaneCallback))
  }

  fun showMovePane(star: Star, fleetId: Long) {
    showStarfield(true /* visible */)

    // TODO: the cast seems... not great.
    val starfieldManager = (context as MainActivity).starfieldManager
    val moveBottomPane = MoveBottomPane(context, starfieldManager) { showActionsPane() }
    moveBottomPane.setFleet(star, fleetId)
    TransitionManager.beginDelayedTransition(bottomPane)
    bottomPane.removeAllViews()
    bottomPane.addView(moveBottomPane)
  }

  fun showSplitPane(star: Star, fleetId: Long) {
    showStarfield(false /* visible */)
    val splitBottomPane = SplitBottomPane(context) { showActionsPane() }
    splitBottomPane.setFleet(star, fleetId)
    TransitionManager.beginDelayedTransition(bottomPane)
    bottomPane.removeAllViews()
    bottomPane.addView(splitBottomPane)
  }

  fun showMergePane(star: Star, fleetId: Long) {
    showStarfield(false /* visible */)
    val mergeBottomPane = MergeBottomPane(context, adapter) { showActionsPane() }
    mergeBottomPane.setFleet(star, fleetId)
    TransitionManager.beginDelayedTransition(bottomPane)
    bottomPane.removeAllViews()
    bottomPane.addView(mergeBottomPane)
  }

  fun showStarfield(visible: Boolean) {
    if (visible) {
      listView.visibility = View.GONE
      background = null
    } else {
      listView.visibility = View.VISIBLE
      setBackgroundResource(R.color.default_background)
    }
  }

  private val actionBottomPaneCallback: ActionBottomPane.Callback = object : ActionBottomPane.Callback {
    override fun onMoveClick() {
      val fleetId = adapter.selectedFleetId ?: return
      val star = adapter.selectedStar ?: return
      showMovePane(star, fleetId)
    }

    override fun onSplitClick() {
      val fleetId = adapter.selectedFleetId ?: return
      val star = adapter.selectedStar ?: return
      showSplitPane(star, fleetId)
    }

    override fun onMergeClick() {
      val fleetId = adapter.selectedFleetId ?: return
      val star = adapter.selectedStar ?: return
      showMergePane(star, fleetId)
    }
  }

  init {
    View.inflate(context, R.layout.fleets, this)
    this.starCollection = starCollection
    bottomPane = findViewById(R.id.bottom_pane)
    listView = findViewById(R.id.fleet_list)
    listView.setAdapter(adapter)
    if (starCollection!!.size() == 1) {
      // if it's just one star, just expand it now.
      listView.expandGroup(0)
    }
    listView.setOnChildClickListener { lv: ExpandableListView?, v: View?, groupPosition: Int, childPosition: Int, id: Long ->
      adapter.onItemClick(groupPosition, childPosition)
      false
    }

    // Actions pane by default.
    bottomPane.addView(ActionBottomPane(getContext(), actionBottomPaneCallback))
  }

}